restoration: checkpoint local allocator state, utilize on restoration

squeed opened the pull request:
This is a stepping stone for the broader [fqdn refactor](https://github.com/cilium/design-cfps/pull/17).

There are 3 changes:
1. Checkpoint local allocator state
2. Use this to generate *all* labels for a prefix upon restart
3. Reduce identity restoration grace period to 30 seconds.

Right now, selector / identity stability is ensured by the ipcache restoration mechanism, which dumps the outgoing ipcache, then requests the same numeric identity for any new prefixes. This is to prevent policy drops as endpoints are regenerated for the first time, as they may select "old" identities until regeneration is complete. Therefore, the goal is that a selector has a consistent set of identities across restarts, and prefixes have a consistent identity.

The current mechanism is simplistic: we dump the outgoing ipcache, list all prefixes, which request their previous numeric identity upon restoration, and we *hope* that their set of labels after restart is the same. This will not work, however, once a single local identity can have multiple prefixes.

#### Example

Imagine a future where we have 3 identities and 3 IPs, all with ID3:
- ID1: `(fqdn:a.com)`
- ID2: `(fqdn:b.com)`
- ID3: `(fqdn:a.com, fqdn:b.com)`

Some endpoints will select (ID1, ID3) and some will select (ID2, ID3). As such, their bpf policy maps will contain these two identities.

The problem comes when we restart the agent. If we do not restore FQDN state exactly (which happens commonly enough), and an IP changes from (a.com, b.com) to (a.com), it could be the case that ID3 now has the set of labels `(fqdn:a.com)`. This could cause policy drops during restoration.

So, a better approach is to restore identities directly, and only fall back to requesting old numeric identities as a fallback.

squeed committed:
go.mod: use json-iterator directly

Previously, it was a transitive dependency.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
identity/cache: checkpoint & restore local allocator state

This writes the state of the local identity allocators to disk,
asynchronously, so that it is available for use during agent restarts.
Checkpoints happen, at most, every 10 seconds, triggered by identity
allocation. JSON streaming is used to reduce memory allocations.

Right now, local identities / labels are restored using heuristics to
try and reconstruct the set of labels for each prefix. Rather than
falling back to heuristics, we should just write out the allocator state
directly.

A subsequent commit will tie this in to the daemon identity restoration
logic.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
daemon: use local allocator checkpoint on restore

This uses the local allocator checkpoint to ensure that prefixes get
identical labels on restart. As before, it dumps the ipache, but now it
reconstructs the set of labels identically.

In the case where the identity is not present in the allocator cache, it
requests the same numeric identity instead.

This allows us to remove a tiny bit of logic in
`IPCache.resolveIdentity()` that was needed to reconstruct node
identities; this heuristic is no longer needed with the checkpoint.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
defaults: reduce identity-restore-grace-period to 30 seconds for k8s

For k8s-only agents, we will never release restored identities until
after k8s caches have synced. So, there's no point in such a long grace
period. Since we now restore all labels, this is a window in which
external CIDRs could potentially have stale labels, so we should keep
this as short as possible.

So, reduce the default to 30 seconds for agents not using kvstore.

However, we have no such waiting period for kvstore-based workloads, so
we must still wait the full 10 minutes.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

derailed commented on daemon/cmd/daemon_main.go:
```diff
@@ -1780,10 +1780,10 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
 		ms.RemoveDisabledMaps()
 
-		// Sleep for the --identity-restore-grace-period (default 10 minutes), allowing
+		// Sleep for the --identity-restore-grace-period (default: 30 seconds k8s, 10 mintues kvstore), allowing
```
nit: typo

derailed commented on daemon/cmd/ipcache.go:
```diff
@@ -102,91 +101,163 @@ func containsSubnet(outer, inner net.IPNet) bool {
+// dumpOldIPache reads the soon-to-be-overwritten ipcache BPF map, noting any prefixes
+// with a locally-scoped or ingress identity.
+func (d *Daemon) dumpOldIPCache() (localPrefixes map[netip.Prefix]identity.NumericIdentity, err error) {
+	localPrefixes = map[netip.Prefix]identity.NumericIdentity{}
```
Does this call warrants named returns if we need to allocate the map here?

squeed replied:
fair enough.

gandro commented on daemon/cmd/ipcache.go:
```diff
@@ -102,91 +101,163 @@ func containsSubnet(outer, inner net.IPNet) bool {
-			d.restoredCIDRs[k.Prefix()] = nid
-		} else if nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero() {
-			ingressIPs = append(ingressIPs, k.Prefix())
+		if nid.Scope() == identity.IdentityScopeLocal || (nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero()) {
```
Why do we no longer use `isLocalIdentity` here? Was the previous check for `IdentityScopeRemoteNode` to preserve the numeric identity, but not the IPCache entry?

squeed replied:
Good question!

There's no longer any point in restoring node entries directly in to the ipcache, since we'll just pick up the same numeric ID when re-upserting the nodes in from k8s / kvstore. In fact, doing so is kind of a bad idea, as IPs could "stick" as being node IPs for up to 10 minutes incorrectly.

gandro commented on pkg/identity/cache/allocator.go:
```diff
@@ -399,15 +434,158 @@ func (m *CachingIdentityAllocator) UnwithholdLocalIdentities(nids []identity.Num
+	// create an immediately-cancelled Context for use during ID restoration.
+	// We will use this to ensure we never do remote identity allocation.
+	closedCtx, cfunc := context.WithCancel(context.Background())
+	cfunc()
```
This doesn't spark joy. I understand that changing the allocator interface would cause lots of churn. 

Since caller and callee are in the same struct, is there no way we move the "allocate local identity" part out of `AllocateIdentity` and have it be a private helper called both from here and `AllocateIdentity`?

squeed replied:
You're right about this for sure. Let me think about it.

To be honest, we ultimately want to change the interface anyways. `ipcache.resolveIdentity()` should never allocate non-local identities anyways, so we need the option. Let me add another method.

squeed replied:
I added `AllocateLocalIdentity()`, which we will need regardless.

gandro reviewed (commented):
Overall looks excellent to me. Two smaller questions/nits

squeed committed:
identity: add AllocateLocalIdentity() to restrict global allocation

This is useful for circumstances in which we *know* we never want to
allocate a global identity, such as the ipcache for CIDR prefixes.

A subsequent commit will make use of this functionality.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

learnitall reviewed (approved):
Vendor changes LGTM!

gandro commented on pkg/identity/cache/allocator.go:
```diff
@@ -286,67 +286,93 @@ func (m *CachingIdentityAllocator) WaitForInitialGlobalIdentities(ctx context.Co
-				} else if id.ID.HasRemoteNodeScope() {
-					metrics.Identity.WithLabelValues(identity.RemoteNodeIdentityType).Inc()
-				} else if id.ID.IsReservedIdentity() {
-					metrics.Identity.WithLabelValues(identity.ReservedIdentityType).Inc()
```
note to reviewers: This case is no longer present in the new version as `allocated` could never become `true` for reserved identities

gandro reviewed (approved):
The new local identity method looks good to me. There is still one unresolved comment from the previous review - but other than that I'm approving.

gandro reviewed (approved):
New changes look straightforward!

gandro commented on daemon/cmd/ipcache.go:
```diff
@@ -102,91 +101,176 @@ func containsSubnet(outer, inner net.IPNet) bool {
+					found = true
+					break
+				}
 			}
```
Nit: I'm assuming you are aware of this, but this algorithmically quadratic in terms of prefixes, as `id == nil` will be true for every prefix during upgrades.

Assuming we're restoring a worst case of 1000 restored prefixes (from the DNS subsystem) I'd expect this to take ~20 milliseconds (1M iterations, based on my recent map-of-prefixes benchmark) upon upgrade (but only upgrade, not regular agent restart). Probably in the range of acceptable. 

However, I do wonder if it makes sense to assemble a quick `counter.Counter[netip.prefix]` counting the identities per prefix and check that one here. In any case, not a blocker for me.

squeed replied:
Good point. Fixed this, and made the code slightly more readable without changing the logic at all.

gandro reviewed (approved):
Manually tested in a local branch. Works like a charm!

derailed reviewed (approved):
@squeed Nice work!
